pkg/assembler/backends/inmem/vulnMetadata.go

```diff
@@ -0,0 +1,302 @@
+	timestamp       time.Time
+	origin          string
+	collector       string
+}
+
+func (n *vulnerabilityMetadataLink) ID() uint32 { return n.id }
+
+func (n *vulnerabilityMetadataLink) Neighbors(allowedEdges edgeMap) []uint32 {
+	out := make([]uint32, 0, 1)
+	if allowedEdges[model.EdgeVulnMetadataVulnerability] {
+		out = append(out, n.vulnerabilityID)
+	}
+	return out
+}
+
+func (n *vulnerabilityMetadataLink) BuildModelNode(c *demoClient) (model.Node, error) {
+	return c.buildVulnerabilityMetadata(n, nil, true)
+}
+
+// Ingest CertifyVuln
+func (c *demoClient) IngestVulnerabilityMetadatas(ctx context.Context, vulnerabilities []*model.VulnerabilityInputSpec, vulnerabilityMetadatas []*model.VulnerabilityMetadataInputSpec) ([]string, error) {
+	// TODO (pxp928): move checks to resolver so all backends don't have to implement
+	if len(vulnerabilities) != len(vulnerabilityMetadatas) {
+		return nil, gqlerror.Errorf("uneven vulnerabilities and vulnerabilityMetadatas for ingestion")
+	}
+	var modelVulnMetadataIDList []string
+	for i := range vulnerabilityMetadatas {
+		vulnMetadata, err := c.IngestVulnerabilityMetadata(ctx, *vulnerabilities[i], *vulnerabilityMetadatas[i])
+		if err != nil {
+			return nil, gqlerror.Errorf("IngestVulnerabilityMetadata failed with err: %v", err)
+		}
+		modelVulnMetadataIDList = append(modelVulnMetadataIDList, vulnMetadata)
+	}
+	return modelVulnMetadataIDList, nil
+}
+
+func (c *demoClient) IngestVulnerabilityMetadata(ctx context.Context, vulnerability model.VulnerabilityInputSpec, vulnerabilityMetadata model.VulnerabilityMetadataInputSpec) (string, error) {
+	return c.ingestVulnerabilityMetadata(ctx, vulnerability, vulnerabilityMetadata, true)
+}
+
+func (c *demoClient) ingestVulnerabilityMetadata(ctx context.Context, vulnerability model.VulnerabilityInputSpec, vulnerabilityMetadata model.VulnerabilityMetadataInputSpec, readOnly bool) (string, error) {
+	funcName := "IngestVulnerabilityMetadata"
+	lock(&c.m, readOnly)
+	defer unlock(&c.m, readOnly)
+
+	var vulnerabilityLinks []uint32
+
+	vulnID, err := getVulnerabilityIDFromInput(c, vulnerability)
+	if err != nil {
+		return "", gqlerror.Errorf("%v ::  %s", funcName, err)
+	}
+	foundVulnNode, err := byID[*vulnIDNode](vulnID, c)
+	if err != nil {
+		return "", gqlerror.Errorf("%v ::  %s", funcName, err)
+	}
+	vulnerabilityLinks = foundVulnNode.vulnMetadataLinks
+
+	var searchIDs []uint32 = vulnerabilityLinks
+
+	// Don't insert duplicates
+	duplicate := false
+	collectedVulnMetadataLink := vulnerabilityMetadataLink{}
+	for _, id := range searchIDs {
+		v, err := byID[*vulnerabilityMetadataLink](id, c)
+		if err != nil {
+			return "", gqlerror.Errorf("%v ::  %s", funcName, err)
+		}
+		vulnMatch := false
+		if vulnID != 0 && vulnID == v.vulnerabilityID {
+			vulnMatch = true
+		}
+		if vulnMatch && vulnerabilityMetadata.Timestamp.Equal(v.timestamp) && vulnerabilityMetadata.ScoreType == v.scoreType &&
+			vulnerabilityMetadata.ScoreValue == float64(v.scoreValue) &&
```

Comment:
Float comparisons should use epsilon helper funcs.

Reply:
Ah good call, something like this?

```
const epsilon = 1e-9
func equal(a, b float64) bool {
    return math.Abs(a - b) <= epsilon
}
```

Reply:
Sorry, like this:
https://github.com/guacsec/guac/blob/main/pkg/assembler/backends/inmem/certifyScorecard.go#L106
https://github.com/guacsec/guac/blob/main/pkg/assembler/backends/inmem/backend.go#L177-L179

Reply:
ah fixed!